ballerina/byte_reader.bal

```diff
@@ -0,0 +1,138 @@
+        ByteIterator temp = self.copybookIterator;
+        self.copybookIterator = self.getIteratorForItem(groupItem);
+
+        if isArray(groupItem) {
+            GroupValue[] elements = [];
+            foreach int i in 0 ..< groupItem.getElementCount() {
+                GroupValue groupValue = {};
+                foreach var child in groupItem.getChildren() {
+                    child.accept(self, groupValue);
+                }
+                elements.push(groupValue);
+            }
+            self.addValue(groupItem.getName(), elements, data);
+        } else {
+            GroupValue groupValue = {};
+            foreach var child in groupItem.getChildren() {
+                child.accept(self, groupValue);
+            }
+            self.addValue(groupItem.getName(), groupValue, data);
+        }
+
+        // Reset the iterator to previous text iterator
+        self.copybookIterator = temp;
+    }
+
+    isolated function visitDataItem(DataItem dataItem, anydata data = ()) {
+        ByteIterator temp = self.copybookIterator;
+        self.copybookIterator = self.getIteratorForItem(dataItem);
+        if isArray(dataItem) {
+            string[] elements = [];
+            foreach int i in 0 ..< dataItem.getElementCount() {
+                elements.push(self.read(dataItem));
+            }
+            self.addValue(dataItem.getName(), elements, data);
+        } else {
+            self.addValue(dataItem.getName(), self.read(dataItem), data);
+        }
+
+        // Reset the iterator to previous text iterator
+        self.copybookIterator = temp;
+    }
+
+    private isolated function getIteratorForItem(DataItem|GroupItem item) returns ByteIterator {
+        string? redefinedItemName = ();
+        if item is GroupItem {
+            redefinedItemName = item.getRedefinedItemName();
+        }
+        if item is DataItem {
+            redefinedItemName = item.getRedefinedItemName();
+        }
+        if redefinedItemName is string {
+            // Obtain the iterator from redfinedValues map if the provided item is a redefining item
+            ByteIterator iterator = new (self.redfinedValues.get(redefinedItemName).toBytes());
+            return iterator;
+        }
+        return self.copybookIterator;
+    }
+
+    private isolated function read(DataItem dataItem) returns string {
+        byte[] bytes = [];
+        int readLength = dataItem.isBinary() ? dataItem.getPackLength() : dataItem.getReadLength();
+        foreach int i in 0 ..< readLength {
+            var data = self.copybookIterator.next();
+            if data is () {
+                break;
+            }
+            bytes.push(data.value);
+        }
+        if bytes.length() == 0 {
+            return "";
+        }
+        if dataItem.isBinary() {
+            int intValue = checkpanic decodeBinaryValue(bytes);
+            return intValue.toString();
+        }
+        return checkpanic string:fromBytes(bytes);
```

Comment:
Any reason to use `checkpanic` here (as it is an anti-pattern) ? If there is no specific usage you could have done this;
```ballerina
return string:fromBytes(bytes);
```

Reply:
Fixed in https://github.com/ballerina-platform/module-ballerinax-copybook/pull/41/commits/4b381aea099f2aecece41287bba54f2f410653ec